src/core/util/latent_see.h

```diff
@@ -0,0 +1,213 @@
+
+  GPR_ATTRIBUTE_ALWAYS_INLINE_FUNCTION static void Append(
+      const Metadata* metadata, EventType type, uint64_t id) {
+    thread_events_.push_back(
+        Event{metadata, std::chrono::steady_clock::now(), id, type});
+  }
+
+ private:
+  Log() = default;
+
+  GPR_ATTRIBUTE_ALWAYS_INLINE_FUNCTION static Log& Get() {
+    static Log* log = []() {
+      atexit([] {
+        LOG(INFO) << "Writing latent_see.json in " << get_current_dir_name();
+        FILE* f = fopen("latent_see.json", "w");
+        if (f == nullptr) return;
+        fprintf(f, "%s", log->GenerateJson().c_str());
+        fclose(f);
+      });
+      return new Log();
+    }();
+    return *log;
+  }
+
+  std::string GenerateJson();
+
+  struct Event {
+    const Metadata* metadata;
+    std::chrono::steady_clock::time_point timestamp;
+    uint64_t id;
+    EventType type;
+  };
+  struct RecordedEvent {
+    uint64_t thread_id;
+    uint64_t batch_id;
+    Event event;
+  };
+  std::atomic<uint64_t> next_thread_id_{1};
+  std::atomic<uint64_t> next_batch_id_{1};
+  static thread_local std::vector<Event> thread_events_;
+  static thread_local uint64_t thread_id_;
+  struct Fragment {
+    Mutex mu;
+    std::vector<RecordedEvent> events ABSL_GUARDED_BY(mu);
+  };
+  PerCpu<Fragment> fragments_{PerCpuOptions()};
+};
+
+template <bool kFlush>
+class Scope {
+ public:
+  GPR_ATTRIBUTE_ALWAYS_INLINE_FUNCTION explicit Scope(const Metadata* metadata)
+      : metadata_(metadata) {
+    Log::Append(metadata_, EventType::kBegin, 0);
+  }
+  GPR_ATTRIBUTE_ALWAYS_INLINE_FUNCTION ~Scope() {
+    Log::Append(metadata_, EventType::kEnd, 0);
+    if (kFlush) Log::FlushThreadLog();
+  }
+
+  Scope(const Scope&) = delete;
+  Scope& operator=(const Scope&) = delete;
+
+ private:
+  const Metadata* const metadata_;
+};
+
+using ParentScope = Scope<true>;
+using InnerScope = Scope<false>;
+
+class Flow {
+ public:
+  GPR_ATTRIBUTE_ALWAYS_INLINE_FUNCTION Flow() : metadata_(nullptr) {}
+  GPR_ATTRIBUTE_ALWAYS_INLINE_FUNCTION explicit Flow(const Metadata* metadata)
+      : metadata_(metadata) {
+    Log::Append(metadata_, EventType::kFlowStart, id_);
+  }
+  GPR_ATTRIBUTE_ALWAYS_INLINE_FUNCTION ~Flow() {
+    if (metadata_ != nullptr) {
+      Log::Append(metadata_, EventType::kFlowEnd, id_);
+    }
+  }
+
+  Flow(const Flow&) = delete;
+  Flow& operator=(const Flow&) = delete;
+  Flow(Flow&& other) noexcept
+      : metadata_(std::exchange(other.metadata_, nullptr)), id_(other.id_) {}
+  Flow& operator=(Flow&& other) noexcept {
+    if (metadata_ != nullptr) Log::Append(metadata_, EventType::kFlowEnd, id_);
+    metadata_ = std::exchange(other.metadata_, nullptr);
+    id_ = other.id_;
+    return *this;
+  }
+
+  GPR_ATTRIBUTE_ALWAYS_INLINE_FUNCTION bool has_value() const {
+    return metadata_ != nullptr;
+  }
+  GPR_ATTRIBUTE_ALWAYS_INLINE_FUNCTION void reset() { emplace(nullptr); }
+  GPR_ATTRIBUTE_ALWAYS_INLINE_FUNCTION void emplace(const Metadata* metadata) {
```

Comment:
Why reinvent optional? optional::emplace will call the destructor of the old instance, and constructor for the new instance, which is effectively what this method does (but maybe in reverse order?)

Reply:
It also costs twice as much memory, which will tend to perturb results more than using less.